src/lib.rs

```diff
@@ -0,0 +1,187 @@
+use crate::config::Config;
+use anyhow::Result;
+use clap::Clap;
+use once_cell::sync::Lazy;
+use regex::Regex;
+
+pub mod config;
+pub mod run;
+
+/// This tool allow you to create a new Rust temporary project in a temporary
+/// directory.
+///
+/// The dependencies can be provided in arguments (e.g.`cargo-temp anyhow
+/// tokio`). When the shell is exited, the temporary directory is deleted unless
+/// you removed the file `TO_DELETE`
+#[derive(Clap, Debug)]
+pub struct Cli {
+    /// Dependencies to add to `Cargo.toml`.
+    ///
+    /// The default version used is `*` but this can be replaced using `=`.
+    /// E.g. `cargo-temp anyhow=1.0.13`
+    #[clap(parse(from_str = parse_dependency))]
+    pub dependencies: Vec<Dependency>,
+
+    /// Create a library instead of a binary.
+    #[clap(long)]
+    pub lib: bool,
+
+    /// Name of the temporary crate.
+    #[clap(long = "name")]
+    pub project_name: Option<String>,
+
+    /// Create a temporary Git working tree based on the repository in the
+    /// current directory
+    #[clap(long = "worktree")]
+    pub worktree_branch: Option<Option<String>>,
+}
+
+impl Cli {
+    pub fn run(&self, config: &Config) -> Result<()> {
+        let tmp_dir = run::generate_tmp_project(
+            self.worktree_branch.clone(),
+            self.project_name.clone(),
+            self.lib,
+            config.temporary_project_dir.clone(),
+        )?;
+
+        run::add_dependencies_to_project(tmp_dir.path(), &self.dependencies)?;
+
+        let delete_file = run::generate_delete_file(tmp_dir.path())?;
+
+        run::start_shell(config, tmp_dir.path())?;
+
+        run::clean_up(delete_file, tmp_dir, self.worktree_branch.clone())?;
+
+        Ok(())
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum Dependency {
+    CrateIo {
+        name: String,
+        version: Option<String>,
+    },
+    Repository {
+        branch: Option<String>,
+        name: String,
+        rev: Option<String>,
+        url: String,
+    },
+}
+
+fn parse_dependency(s: &str) -> Dependency {
+    // This will change when `std::lazy` is released.
+    // See https://github.com/rust-lang/rust/issues/74465.
+    static RE: Lazy<Regex> = Lazy::new(|| {
+        Regex::new(r"^(?P<name>[^=]+)=(?P<version>((?P<url>\w+://([^:@]+(:[^@]+)?@)?[^#]+)(#branch=(?P<branch>.+)|#rev=(?P<rev>.+))?)|.+)$")
+            .unwrap()
+    });
+
+    if let Some(caps) = RE.captures(s) {
+        if let Some(url) = caps.name("url") {
+            Dependency::Repository {
+                name: caps
+                    .name("name")
+                    .map(|x| x.as_str().to_string())
+                    .expect("Cannot get repository name"),
```

Comment:
No reason for this to fail.
```suggestion
                name: caps.name("name").unwrap().as_str().to_string(),
```